src/slide-image-source.cc

```diff
@@ -0,0 +1,164 @@
+// -*- mode: c++; c-basic-offset: 4; indent-tabs-mode: nil; -*-
+// (c) 2021 Leonardo Romor <leonardo.romor@gmail.com>,
+//          Henner Zeller <h.zeller@acm.org>
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation version 2.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://gnu.org/licenses/gpl-2.0.txt>
+
+#include "slide-image-source.h"
+
+#include "framebuffer.h"
+#include "terminal-canvas.h"
+
+#include <cassert>
+#include <cstdio>
+#include <unistd.h>
+
+#include <limits>
+#include <string>
+#include <utility>
+
+#include <openslide.h>
+
+extern "C" {
+#  include <libswscale/swscale.h>
+}
+
+namespace timg {
+const char *SlideImageSource::VersionInfo() {
+    return openslide_get_version();
+}
+
+static void dummy_log(void *, int, const char *, va_list) {}
+
+// Scoped clean-up of c objects.
+struct ScopeGuard {
+    ScopeGuard(std::function<void()> f) : f_(f) {}
+    ~ScopeGuard() { f_(); }
+    std::function<void()> f_;
+};
+
+static inline bool invalid_dimensions(int64_t width, int64_t height) {
+    return (width < 0 || height < 0) ||
+        (width > std::numeric_limits<int>::max() ||
+         height > std::numeric_limits<int>::max());
+}
+
+bool SlideImageSource::LoadAndScale(const DisplayOptions &opts, int, int) {
+    options_ = opts;
+    if (opts.scroll_animation ||
+        filename() == "/dev/stdin" || filename() == "-") {
+        return false;  // Not dealing with these now.
+    }
+
+    // Check if this format is supported
+    openslide_t *osr = openslide_open(filename().c_str());
+    if (!osr)
+        return false;
+
+    // Automatically cleanup openslide context
+    ScopeGuard s([osr]() { openslide_close(osr); });
+
+    std::unique_ptr<timg::Framebuffer> source_image;
+    int64_t width, height, l0_width, l0_height;
```

Comment:
The lower-case `l` is hard to distinguish from the number `1`.
So I suggest to name these `level0_width` and `level0_height`

Reply:
(solved)